spec/02-integration/03-db/03-plugins_spec.lua

```diff
@@ -60,6 +72,23 @@ for _, strategy in helpers.each_strategy() do
                       [[name="key-auth",route={id="]] .. route.id ..
                       [["},consumer=null}']], err_t.message)
         end)
+
+        it("returns an error when inserting mismatched plugins", function()
+          local plugin, _, err_t = db.plugins:insert({ name = "key-auth",
+                                                       protocols = { "http" },
+                                                       route = { id = route.id },
+                                                     })
+          assert.is_nil(plugin)
+          assert.equals(err_t.fields.protocols, "Plugin protocols must match the associated route's protocols")
+
+          local plugin, _, err_t = db.plugins:insert({ name = "key-auth",
+                                                       protocols = { "tcp" },
+                                                       service = { id = service.id },
+                                                     })
+          assert.is_nil(plugin)
+          assert.equals(err_t.fields.protocols,
+                        "Plugin protocols must match at least one of the service's route's protocols")
```

Comment:
"service's route's protocols" is a mouthful, I don't think it's usual form to double "'s" like this to "nest" ownership. I suggest `"must match the protocols of at least one route pointing to this Plugin's service"`.

To confirm, the intent of this verification is for user-friendliness only and not actual _validation_, right? Because if I follow correctly, I can add `key-auth` to route `R` with { protocols = {"http"} }` and then PATCH `R` to `{ protocols = {"tcp"} }` and the plugin will remain plugged to it, breaking the invariant that plugins are only added to routes of the correct protocol.

I mean, there is no actual _harm_ if that happens right? (Other than having a plugin attached to route `R` that will never be triggered) — I'm not advocating to add checks in the Routes and Services DAOs to verify that the constraint is never broken, just trying to understand whether we really care if the invariant is held or not.

Reply:
I updated the message as requested.

> To confirm, the intent of this verification is for user-friendliness only and not actual validation, right?

Yes, the compatibility errors are 100% UX.

> I mean, there is no actual harm if that happens right?

Nothing bad should happen.

* First of all, the plugin should not run on the same Lua space where the route is matched, given the first `perf` commit on this PR, which filters them out.
* Secondly, even if they somehow managed to run together, stream plugins have different phases than http plugins.